### PR TITLE
Add nGraph as a git submodule (#308)

### PR DESCRIPTION
**Pull request by cconvey**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -399,7 +382,12 @@ SRC = $(wildcard src/*/*/*/*.cc src/*/*/*.cc src/*/*.cc src/*.cc)
 ifneq ($(USE_NGRAPH),1)
     SRC := $(foreach f,$(SRC),$(if $(findstring src/ngraph,$f),,$f))
 endif
+
 OBJ = $(patsubst %.cc, build/%.o, $(SRC))
+ifeq ($(USE_NGRAPH),1)
+  $(OBJ): ngraph
```
Comment (mbrookhart) [inline Makefile]: I'm not familiar with this syntax, what does this line do?

File: NGRAPH_README.md [hunk 2]
```diff
@@ -1,4 +1,20 @@
...
+  without any dependency on, or support for, nGraph.
+- If `USE_NGRAPH` is set to `1` but `NGRAPH_DIR` is undefined or blank, then MXnet's build
+  system will build and use its own copy of nGraph, as found under the `3rdparty`
+  source directory.
+- If `USE_NGRAPH` is set to `1` and `NGRAPH_DIR` is set to a non-empty string, then MXnet's
+  build system will use an already-installed build of nGraph, residing under the `NGRAPH_DIR`
+  directory.
+
+When building MXnet with nGraph support, several additional nGraph-specific Make variables
+come into play.  Please see `make/config.mk` for more details.
```
Comment (mbrookhart) [inline NGRAPH_README.md]: 👍

File: make/config.mk [hunk 3]
```diff
@@ -107,7 +107,15 @@ USE_NGRAPH_DISTRIBUTED = 0
 # e.g. NGRAPH_DIR = <path>/ngraph/install (containing lib and include dirs)
 # if not specified nGraph will be cloned/built/installed during make
 NGRAPH_DIR =
-
+# If USE_NGRAPH=1 and NGRAPH_DIR is undefined/empty, then the following variables
+# are used to configure/build a local copy of nGraph:
+NGRAPH_EXTRA_MAKE_FLAGS="VERBOSE=1"
+NGRAPH_EXTRA_CMAKE_FLAGS="-DNGRAPH_USE_PREBUILT_LLVM=1"
```
Comment (mbrookhart) [inline make/config.mk]: This isn't valid on some systems (Mac, Centos). I guess we need more complicated logic here?

File: ngraph.mk [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  ifeq ($(USE_MKLDNN), 1)
+    $(error "Cannot have both USE_NGRAPH=1 and USE_MKLDNN=1: they require different MKLDNN versions.")
+  endif
+endif
+
+#===================================================================================================
+
+ifeq ($(USE_NGRAPH), 1)
+  ifeq ($(NGRAPH_DIR),)
+    NGRAPH_INSTALL_DIR=$(ROOTDIR)/3rdparty/ngraph.install
```
Comment (mbrookhart) [inline ngraph.mk]: ngraph.install or ngraph/install?

File: ngraph.mk [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  else
+    NGRAPH_INSTALL_DIR=$(NGRAPH_DIR)
+  endif
+
+  NGRAPH_LIB_DIR=$(NGRAPH_INSTALL_DIR)/lib
+  NGRAPH_INCLUDE_DIR=$(NGRAPH_INSTALL_DIR)/include
+
+.PHONY: ngraph_clean
+
+  ifeq ($(NGRAPH_DIR),)
```
Comment (mbrookhart) [inline ngraph.mk]: The tabbing on this file is a little confusing, I guess that's due to Make's tab requirements?

File: Makefile (hunk 1 above)
Comment (cconvey) [reply Makefile]: `$(OBJ)` expands to a list of `.o` files.  This states that every one of those `.o` files is dependent on the `ngraph` target first being built.

This was the cleanest way I could think of to ensure that ngraph was fully built and installed before any of the code that uses it.

File: make/config.mk (hunk 3 above)
Comment (cconvey) [reply make/config.mk]: AFAIK the only currently supported platform for nGraph is Ubuntu 16.04.  Is that incorrect?

File: ngraph.mk (hunk 4 above)
Comment (cconvey) [reply ngraph.mk]: Could go either way.  In general I try to avoid mixing source, build, and install directory trees.  I can change this if you feel strongly.

File: ngraph.mk (hunk 5 above)
Comment (cconvey) [reply ngraph.mk]: Unfortunately yeah.  For any rules, the first line must have no indentation, and the recipe lines must be intended by (at least) one leading tab character.

The other statements in a Makefile, however, are free to use indentation however they want (AFAIK).

Because I had several layers of `if` nesting, I thought the lesser of two evils was to indent nested `if`s.

**Comment by cconvey** [comment]
> If the user isn't specifying an NGRAPH_DIR, we should copy NGRAPH_INSTALL_DIR/lib/* to MXNET_LIBDIR so the user doesn't need to set LD_LIBRARY_PATH

Yeah, I can do this.  I'm not sure (either way) how well that approach will hold up when I tackle #307, but we can cross that bridge when we come to it.

IMO the better solution is for nGraph's build system offer us a way to specify precisely which directory should contain the libraries produced by its `install` target.

File: Makefile [hunk 6]
```diff
@@ -78,6 +78,8 @@ ifeq ($(USE_MKLDNN), 1)
 	export USE_MKLML = 1
 endif
 
+include ngraph.mk
```
[the remark this answers is not in the document]
Comment (cconvey) [reply Makefile]: We can, but I was trying to mostly match the style used by the DO guys, who unconditionally `include mkldnn.mk`.

I'm not passionate either way, so whatever you want is fine.

File: ngraph.mk (hunk 4 above)
Comment (mbrookhart) [reply ngraph.mk]: Nope, just wanted to make sure it was intentional.

File: make/config.mk (hunk 3 above)
Comment (mbrookhart) [reply make/config.mk]: We also officially support Centos 7 and Mac is on the priority list this quarter.

File: ngraph.mk [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  ifeq ($(USE_NGRAPH_DISTRIBUTED), 1)
+    MPI_LINK_FLAGS = $(shell mpicxx --showme:link)
+    NGRAPH_LDFLAGS += $(MPI_LINK_FLAGS)
+  endif
+else
+  #-------------------------------------------------------------------------------------------------
+  # USE_NGRAPH != 1
+  #-------------------------------------------------------------------------------------------------
+  NGRAPH_CFLAGS=
+  NGRAPH_LDFLAGS=
```
Comment (mbrookhart) [inline ngraph.mk]: if nGraph != 1, this will fail without an ngraph target?

File: ngraph.mk [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	$(RM) -r "$(NGRAPH_BUILD_DIR)"
+	$(RM) -r "$(NGRAPH_INSTALL_DIR)"
+
+ifeq ($(USE_NGRAPH), 1)
+
+.PHONY: ngraph
+all: ngraph
+ngraph:
+	m
+	kdir -p "$(NGRAPH_BUILD_DIR)"
```
Comment (mbrookhart) [inline ngraph.mk]: broken line split

File: ngraph.mk [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+NGRAPH_SRC_DIR := $(ROOTDIR)/3rdparty/ngraph
+NGRAPH_BUILD_DIR := $(ROOTDIR)/3rdparty/ngraph/build
+NGRAPH_INSTALL_DIR := $(ROOTDIR)/3rdparty/ngraph/install
+MXNET_LIB_DIR := $(ROOTDIR)/lib
+
+# The 'clean' target should remove nGraph-related generated files, regardless of whether or not
+# the current Make invocation has USE_NGRAPH=1 ...
+.PHONY: ngraph_clean
+clean: ngraph_clean
```
Comment (mbrookhart) [inline ngraph.mk]: I don't think you need this, the dependency is defined in the Makefile?

File: ngraph.mk (hunk 7 above)
Comment (mbrookhart) [reply ngraph.mk]: i.e.:
```
-> % make -j56 USE_NGRAPH=0
Makefile:295: WARNING: Significant performance increases can be achieved by installing and enabling gperftools or jemalloc development packages
make: *** No rule to make target 'build/src/operator/nn/mkldnn/mkldnn_convolution.o', needed by 'lib/libmxnet.a'.  Stop.
```

**Comment by lnguyen-nvn** [comment]
TESTNOW

**Comment by lnguyen-nvn** [comment]
TESTNOW

**Comment by lnguyen-nvn** [comment]
TESTNOW

**Comment by lnguyen-nvn** [comment]
TESTNOW

**Comment by nlsapp** [comment]
TESTNOW

**Comment by nlsapp** [comment]
TESTNOW

**Comment by nlsapp** [comment]
TESTNOW

**Comment by lnguyen-nvn** [comment]
TESTNOW

File: python/mxnet/libinfo.py [hunk 10]
```diff
@@ -74,6 +131,13 @@ def find_lib_path():
                            'List of candidates:\n' + str('\n'.join(dll_path)))
     if os.name == 'nt':
         os.environ['PATH'] = os.environ['PATH'] + ';' + os.path.dirname(lib_path[0])
+
+    # We're only intersted in the entry at the head of the list...
+    lib_path = lib_path[0:1]
+
+    mxnet_lib_dir = os.path.dirname(lib_path[0])
+    lib_path += _get_ngraph_shared_lib_paths(mxnet_lib_dir)
```
Comment (mbrookhart) [inline python/mxnet/libinfo.py]: This will always fail if mxnet is compiled without nGraph, we need a better solution.
